SeQuant/core/space.hpp

```diff
@@ -417,109 +314,446 @@ class IndexSpace {
-    if (!instance_exists(attr)) throw bad_attr();
-    return attr2basekey_.find(attr)->second;
-  }
 
   /// Default ctor creates space with nonnull type and null quantum numbers
-  IndexSpace() noexcept : attr_(nonnulltype, nullqns) {}
+  IndexSpace() {
+    attr_ = {{0x7fffffff}, nullqns};
+    base_key_ =L"";
+    approximate_size_ = 0;
+  }
 
   IndexSpace(const IndexSpace &other) {
     if (!other.attr().is_valid())
       throw std::invalid_argument(
           "IndexSpace copy ctor received invalid argument");
-    attr_ = other.attr_;
+  attr_ = other.get_attr();
+  base_key_ = other.get_base_key();
+  approximate_size_ = other.get_approximate_size();
   }
   IndexSpace(IndexSpace &&other) {
     if (!other.attr().is_valid())
       throw std::invalid_argument(
           "IndexSpace move ctor received invalid argument");
-    attr_ = other.attr_;
+    attr_ = other.get_attr();
+base_key_ = other.get_base_key();
+approximate_size_ = other.get_approximate_size();
   }
   IndexSpace &operator=(const IndexSpace &other) {
     if (!other.attr().is_valid())
       throw std::invalid_argument(
           "IndexSpace copy assignment operator received invalid argument");
-    attr_ = other.attr_;
+    attr_ = other.get_attr();
+    base_key_ = other.get_base_key();
+    approximate_size_ = other.get_approximate_size();
     return *this;
   }
   IndexSpace &operator=(IndexSpace &&other) {
     if (!other.attr().is_valid())
       throw std::invalid_argument(
           "IndexSpace move assignment operator received invalid argument");
-    attr_ = other.attr_;
+    attr_ = other.get_attr();
+    base_key_ = other.get_base_key();
+    approximate_size_ = other.get_approximate_size();
     return *this;
   }
 
-  /// @brief constructs an instance of an IndexSpace object
-  IndexSpace(Type type, QuantumNumbers qns) noexcept : attr_(type, qns) {
-    assert(attr_.is_valid());
+
+  const Attr get_attr()const {
+    return attr_;
+  }
+  std::wstring get_base_key() const{
+    return base_key_;
+  }
+  static std::wstring_view reduce_key(std::wstring_view key) {
+    const auto underscore_position = key.rfind(L'_');
+    if(underscore_position != -1){ // key can be reduced
+      return key.substr(0, underscore_position);
+    }
+    else{
+      return key;
+    }
+  }
+
+  // having approximate sizes of spaces can be helpful when optimizing evaluation order.
+  unsigned long get_approximate_size() const {
+    return approximate_size_;
   }
 
  private:
   Attr attr_ = Attr::invalid();
+  std::wstring base_key_;
+  unsigned long approximate_size_;
   /// @brief constructs an instance of an IndexSpace object
-  explicit IndexSpace(Attr attr) noexcept : attr_(attr) {
-    assert(attr.is_valid());
+  //explicit IndexSpace(Attr attr) noexcept : attr_(attr) {
+  //  assert(attr.is_valid());
+  //}
+
+
+
+  static std::wstring to_wstring(std::wstring_view key) {
+    return std::wstring(key.begin(), key.end());
   }
```

Comment:
What is this used for? It should be completely equivalent to simply writing `std::wstring(key)` - implementing a wrapper function for this doesn't seem to gain much